go/src/koding/kites/kloud/provider/builder.go

```diff
@@ -311,13 +311,33 @@ func (bs *BaseStack) applyAsync(ctx context.Context, req *kloud.ApplyRequest) er
 		return err
 	}
 
+	// At this point resources have been built successfully on the terraform side.
+	// If updating stack or provisioning klient on remote hosts fails,
+	// we need to cleanup the stack; if debug: true was set in the stack template
+	// for any of the instance resources, the stack won't be destroyed
+	// so it's possible to troubleshoot it post-mortem.
+	if !bs.Debug {
+		defer func() {
+			if err != nil {
+				_, e := tfKite.Destroy(&tf.TerraformRequest{
```

Comment:
Terraformer kite does not do incremental updates (if you meant retrying apply if state got partially transitioned to the expected one).

This is best-effort attempt of fixing e.g. leaked vms, when apply partially failed.

At this point I am not sure whether not updating state file on failure was intentional design for terraformer or not. But taking into consideration fact, that we do not store klient connectivity result in a state file [0] the apply was most likely mean to be one-shot. Correct me if I'm wrong.

[0] So the vm resource has a chance to be recreated when klient for whatever reason does not connect to kontrol.